base/array.jl

```diff
@@ -255,14 +255,14 @@ else
 end
 _default_eltype{I,T}(::Type{Generator{I,Type{T}}}) = T
 
-_array_for(T, itr, ::HasLength) = Array(T, Int(length(itr)::Integer))
-_array_for(T, itr, ::HasShape) = Array(T, convert(Dims,size(itr)))
+_array_for(T, itr, ::HasLength) = Array{T,ndims(itr)}(Int(length(itr)::Integer))
+_array_for(T, itr, ::HasShape) = Array{T,ndims(itr)}(convert(Dims,size(itr)))
 
```

Comment:
I might have expressed a bit wrong. The problem is that since `ndims(itr)` is not inferable the compiler cannot inline the constructor where it might help speed something.